Float funque plus features

MallikarjunKamble opened the pull request:
**Floating Point Implementation for new features:**

- Added support for 5-Tap nadeanu spatial csf
- Added support for nadenau_weight, li, hill, watson and mannos_weight wavelet csf
- Added MS_ESSIM feature
- Added ST-RRED feature

**Fixed:**

- Repeated aligned_malloc memory allocation for dwt buffers has been removed
- Added json file for float_funque
- Added features list in integer_funque

cosmin commented on libvmaf/src/feature/third_party/funque/funque_ssim.c:
```diff
@@ -93,5 +93,180 @@ int compute_ssim_funque(dwt2buffers *ref, dwt2buffers *dist, double *score, int
+#if ENABLE_MINK3POOL
+    double ssim_val = 1 - cbrt((double) cube_1minus_map / (width * height));
+    score->ssim_mean = ssim_clip(ssim_val, 0, 1);
+#else
```
I think we can drop all the CoV pooling code and stick with mink pooled only as `essim` and `ms-essim` and simultaneously compute mean pooled values as `ssim` and `ms-ssim`.

MallikarjunKamble replied:
Mink3 pooling is added. Covariance based pooling can be removed after review

cosmin commented on libvmaf/src/feature/third_party/funque/x86/hbd_resizer_avx2.c:
```diff
@@ -346,6 +345,7 @@ void hbd_vresize_avx2(const int **src, unsigned short *dst, const short *beta, i
         __m256i accum_23_0 = _mm256_add_epi64(mul2_0, mul3_0);
         __m256i accum_01_4 = _mm256_add_epi64(mul0_4, mul1_4);
         __m256i accum_23_4 = _mm256_add_epi64(mul2_4, mul3_4);
+
```
can we clean up these unrelated whitespace changes from this pull request and leave only the actually modified files?

MallikarjunKamble commented:
**Fixed Point Implementation:**

- Added support for 5-Tap nadeanu spatial csf
- Added support for wavelet csf (nadenau weights)
- Added MS_SSIM integer feature
- Added ST-RRED integer feature

**Enhancements to Floating point:**

- Memory optimizations in MS_SSIM
- Tested floating point code to mactch across all 4 levels. Added fixes to match the python code.
-  - Fixes added in MS_SSIM and STRRED float code to match all 4 levels

abhinaukumar commented on libvmaf/src/feature/third_party/funque/funque_filters.c:
```diff
@@ -132,58 +131,63 @@ void spatial_filter(float *src, float *dst, int width, int height)
+
+void spatial_csfs(float *src, float *dst, int width, int height, float *tmp_buf, int num_taps)
+{
+    int fwidth;
```
`fwidth` appears to be a duplicate of `num_taps`.

MallikarjunKamble replied:
We will address num_taps redundancy in the next commit

abhinaukumar commented on libvmaf/src/feature/third_party/funque/integer_funque_filters.c:
```diff
@@ -302,18 +319,27 @@ static inline void integer_horizontal_filter(spat_fil_inter_dtype *tmp, spat_fil
+void integer_spatial_filter(void *src, spat_fil_output_dtype *dst, int dst_stride, int width,
+                            int height, int bitdepth, spat_fil_inter_dtype *tmp, int num_taps)
+{
+    int fwidth;
```
`fwidth` appears to be a duplicate of `num_taps`.

MallikarjunKamble replied:
We will address num_taps redundancy in the next commit

abhinaukumar commented on libvmaf/src/feature/third_party/funque/funque_ssim.c:
```diff
@@ -93,5 +93,149 @@ int compute_ssim_funque(dwt2buffers *ref, dwt2buffers *dist, double *score, int
+            cs_sum += cs;
+            ssim_sq_sum += (l * cs) * (l * cs);
+            l_sq_sum += l * l;
+            cs_sq_sum += cs * cs;
```
`l` and `cs` need to be mink3 pooled for use in MS-ESSIM.

MallikarjunKamble replied:
Addressed. Please review and close.

abhinaukumar commented on libvmaf/src/feature/third_party/funque/funque_ssim.c:
```diff
@@ -93,5 +93,149 @@ int compute_ssim_funque(dwt2buffers *ref, dwt2buffers *dist, double *score, int
+    return ret;
+}
+
+int compute_ms_ssim_mean_scales(MsSsimScore* score, int n_levels)
```
This function doesn't appear to be handled appropriately under `ENABLE_MINK3POOL`. Flagging to double-check.

MallikarjunKamble replied:
This is ssim. We will correct the code in ssim and add support for mink3 pooling and update in next commit.

abhinaukumar commented on libvmaf/src/output.c:
```diff
@@ -99,7 +99,7 @@ int vmaf_write_output_xml(VmafContext *vmaf, VmafFeatureCollector *fc,
                                                 0, pic_cnt - 1);
             if (!err)
-                fprintf(outfile, "%s=\"%.6f\" ", pool_method_name[j], score);
+                fprintf(outfile, "%s=\"%.30f\" ", pool_method_name[j], score);
```
Are these `.30f` formattings deliberately left in or are they holdovers from debugging?

MallikarjunKamble replied:
This is to take care of ST-RRED accuracy. ST-RRED values go upto E-30

MallikarjunKamble commented:
**Fixed Point Implementation:**

- Addition of Multi-Scale feature computation (All 4 scales)
- Addition of support for crop width and crop height
- Addition of MINK3 pool in MS_SSIM
- Bug fixes for accuracy improvements wrt python

**Enhancements to Floating point:**

- Addition of support for crop width and crop height
- Addition of MINK3 pool in MS_SSIM
- Addition of filter coeffs for Li and Hill
